### PR TITLE
KeyVault HostingStartup

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/aspnet/DataProtection/issues/279

Not Identity for now but it will follow same pattern.

### CODE REVIEW

File: src/Microsoft.AspNetCore.AzureKeyVault.HostingStartup/Microsoft.AspNetCore.AzureKeyVault.HostingStartup.csproj [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>aspnetcore;keyvault;azure</PackageTags>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.AzureKeyVault" />
+    <PackageReference Include="Microsoft.AspNetCore.DataProtection.AzureKeyVault" />
+    <PackageReference Include="Microsoft.AspNetCore.Identity.Service.AzureKeyVault" />
+    <PackageReference Include="Microsoft.Azure.Services.AppAuthentication" Version="1.0.0-preview" />
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.AzureKeyVault.HostingStartup/Microsoft.AspNetCore.AzureKeyVault.HostingStartup.csproj]: https://github.com/aspnet/Universe/pull/617

**Comment by Tratcher** [comment]
I was under the impression this would be a new site extension rather than being added to the existing one. @glennc ?

File: AzureIntegration.sln [hunk 2]
```diff
@@ -46,9 +46,9 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.AzureA
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.AzureAppServices.TestBundle", "src\Microsoft.AspNetCore.AzureAppServices.TestBundle\Microsoft.AspNetCore.AzureAppServices.TestBundle.csproj", "{1EC31DA1-131D-4257-B001-BE8391E6077E}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Web.Xdt.Extensions.Tests", "test\Microsoft.Web.Xdt.Extensions.Tests\Microsoft.Web.Xdt.Extensions.Tests.csproj", "{809AEE05-1B28-4E31-8959-776B249BD725}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Web.Xdt.Extensions.Tests", "test\Microsoft.Web.Xdt.Extensions.Tests\Microsoft.Web.Xdt.Extensions.Tests.csproj", "{809AEE05-1B28-4E31-8959-776B249BD725}"
```
Comment (Tratcher) [inline AzureIntegration.sln]: ?

File: src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/HostingStartupConfigurationExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            return new ConfigurationBuilder()
+                .AddEnvironmentVariables(prefix: "ASPNETCORE_")
+                .Build();
+        }
+        public static bool IsEnabled(this IConfiguration configuration, string hostingStartupName, string featureName)
+        {
+            if (configuration.TryGetOption(hostingStartupName, featureName, out var value))
+            {
+                value = value.ToLowerInvariant();
+                return value != "false" && value != "0";
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/HostingStartupConfigurationExtensions.cs]: on if not correctly turned off? hmm.

File: test/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests/HostingStartupTests.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+using Xunit;
+
+namespace Microsoft.AspNetCore.AzureKeyVault.HostingStartup.Tests
+{
+    public class HostinStartupTests
+    {
+        [Fact]
+        public void Configure_AddsDataProtection()
+        {
+            Environment.SetEnvironmentVariable("ASPNETCORE_HostingStartup__KeyVault__DataProtection", null);
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests/HostingStartupTests.cs]: HostingStartup doesn't seem important in these values.
`ASPNETCORE__KeyVault__DataProtection`
`ASPNETCORE__KeyVault__DataProtectionKey`
?

File: src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/Microsoft.AspNetCore.AzureAppServices.HostingStartup.csproj [hunk 5]
```diff
@@ -17,4 +17,12 @@
     <ProjectReference Include="..\Microsoft.AspNetCore.AzureAppServicesIntegration\Microsoft.AspNetCore.AzureAppServicesIntegration.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Configuration.EnvironmentVariables" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.AzureKeyVault" />
+    <PackageReference Include="Microsoft.AspNetCore.DataProtection.AzureKeyVault" />
+    <PackageReference Include="Microsoft.AspNetCore.Identity.Service.AzureKeyVault" />
+    <PackageReference Include="Microsoft.Azure.Services.AppAuthentication" Version="1.0.0-preview" />
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/Microsoft.AspNetCore.AzureAppServices.HostingStartup.csproj]: extract version

File: src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/AzureKeyVaultHostingStartup.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                    .Build();
+
+                if (configuration.IsEnabled(HostingStartupName, DataProtectionFeatureName) &&
+                    configuration.TryGetOption(HostingStartupName, DataProtectionKeyName, out var protectionKey))
+                {
+                    AddDataProtection(collection, keyVaultClient, protectionKey);
+                }
+            });
+
+            if (baseConfiguration.IsEnabled(HostingStartupName, ConfigurationFeatureName) &&
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/AzureKeyVaultHostingStartup.cs]: Why do you need two keys for this if you're always going to check both? Why not enable based only on the ConfigurationVaultName? Same for DataProtectionKeyName.

File: AzureIntegration.sln (hunk 2 above)
Comment (pakrym) [reply AzureIntegration.sln]: Will revert.

File: src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/HostingStartupConfigurationExtensions.cs (hunk 3 above)
Comment (pakrym) [reply src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/HostingStartupConfigurationExtensions.cs]: On by default :)

File: src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/Microsoft.AspNetCore.AzureAppServices.HostingStartup.csproj (hunk 5 above)
Comment (pakrym) [reply src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/Microsoft.AspNetCore.AzureAppServices.HostingStartup.csproj]: https://github.com/aspnet/Universe/pull/617

File: test/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests/HostingStartupTests.cs (hunk 4 above)
Comment (pakrym) [reply test/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests/HostingStartupTests.cs]: Didn't want to take nice names and wanted it to show that it's only hosting startup specific.

File: src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/AzureKeyVaultHostingStartup.cs (hunk 6 above)
Comment (pakrym) [reply src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/AzureKeyVaultHostingStartup.cs]: Just to keep a same pattern for enabling features. Some of them wouldn't have parameters and I'd like to have a common way for enabling\disabling.

Comment (Tratcher) [reply src/Microsoft.AspNetCore.AzureAppServices.HostingStartup/AzureKeyVaultHostingStartup.cs]: Can the key name say "ConfigurationEnabled" at least?

**Comment by pakrym** [comment]
:up::date:

File: AzureIntegration.sln [hunk 7]
```diff
@@ -46,10 +46,6 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.AzureA
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.AzureAppServices.TestBundle", "src\Microsoft.AspNetCore.AzureAppServices.TestBundle\Microsoft.AspNetCore.AzureAppServices.TestBundle.csproj", "{1EC31DA1-131D-4257-B001-BE8391E6077E}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Web.Xdt.Extensions.Tests", "test\Microsoft.Web.Xdt.Extensions.Tests\Microsoft.Web.Xdt.Extensions.Tests.csproj", "{809AEE05-1B28-4E31-8959-776B249BD725}"
```
Comment (Tratcher) [inline AzureIntegration.sln]: ?

File: test/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests/HostingStartupTests.cs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            Assert.True(callbackCalled);
+        }
+
+        [Theory]
+        [InlineData("0")]
+        [InlineData("FALSE")]
+        [InlineData("false")]
+        public void Configure_SkipsAddsDataProtection_IfDisabled(string value)
+        {
+            Environment.SetEnvironmentVariable("ASPNETCORE_HostingStartup__KeyVault__DataProtection", value);
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests/HostingStartupTests.cs]: Enabled.
None of the tests were updated with the new key name.

File: test/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests.csproj [hunk 9]
```diff
@@ -7,6 +7,7 @@
 
   <ItemGroup>
     <ProjectReference Include="..\..\src\Microsoft.AspNetCore.AzureAppServicesIntegration\Microsoft.AspNetCore.AzureAppServicesIntegration.csproj" />
+    <ProjectReference Include="..\..\src\Microsoft.AspNetCore.AzureAppServices.HostingStartup\Microsoft.AspNetCore.AzureAppServices.HostingStartup.csproj" />
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests.csproj]: ?

Comment (pakrym) [reply test/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests/Microsoft.AspNetCore.AzureAppServicesIntegration.Tests.csproj]: Didn't add another test project and reuse `Microsoft.AspNetCore.AzureAppServicesIntegration.Tests.csproj` for tests

**Pull request merged**